tithemi/assignment2/Q1.cpp

```diff
@@ -0,0 +1,53 @@
+#include <array>
+#include <vector>
+#include <iostream>
+
+template<typename T, size_t N = 2>
+struct Node
+{
+    T value;
+    std::array<Node<T>*, N> children;
+
+    Node(const T value, std::array<Node<T>*, N> children)
+            : value(value), children(children)
+    { }
+
+    bool find(const T& target, std::vector<Node*>& path)
+    {
+        if (value == target)
+            return true;
+
+        for (Node<T>* child : children)
+        {
+            if (child != nullptr && child->find(target, path))
+            {
+                path.push_back(this);
+                return true;
+            }
+        }
+
+        return false;
+    }
+};
+
+int main()
+{
+    Node<int>* root = new Node<int>(7, {
```

Comment:
One issue with the code in main here and in Q2.cpp is that you're leaking memory.

Every time you create an object using `new` syntax, you are allocating memory that is not cleaned up implicitly. Therefore, for every object a function creates in this way, it must make sure to release the memory occupied by that object when it's done. For that you have three options:

1.  Call `delete` recursively on all of the nodes after you're done with them.

2.  Write a destructor, `~Node()`, deleting all allocated memory per node. The destructor will be automatically called when the object goes out of scope, so it's strictly better than 1.

3. Wrap all pointers in [std::unique_ptr](http://www.drdobbs.com/cpp/c11-uniqueptr/240002708) and never even think about what do to with memory you're allocating, the unique_ptr will figure it out for you. This option is strictly better than 1. and 2. for most cases.

Please change the code here and in Q2.cpp accordingly, to not leak memory :)